rkvst_simplehash/v1.py

```diff
@@ -0,0 +1,92 @@
+""" Module for implementation of simplehash canonicalization"""
+
+from hashlib import sha256
+from json import load as json_load
+from operator import itemgetter
+from sys import stdin as sys_stdin
+
+from bencodepy import encode as binary_encode
+
+V1_FIELDS = [
+    "identity",
+    "asset_identity",
+    "event_attributes",
+    "asset_attributes",
+    "operation",
+    "behaviour",
+    "timestamp_declared",
+    "timestamp_accepted",
+    "timestamp_committed",
+    "principal_accepted",
+    "principal_declared",
+    "confirmation_status",
+    "from",
+    "tenant_identity",
+]
+
+
+class SimpleHashPendingEventFound(Exception):
+    """If PENDING event found"""
+
+
+class SimpleHashFieldMissing(Exception):
+    """If essential field is missing"""
+
+
+def check_for_pending_events(events):
+    """Raise exception if any PENDING events found"""
+
+    pending = next(
+        (
+            event["identity"]
+            for event in events
+            if event["confirmation_status"] == "PENDING"
+        ),
+        None,
+    )
+    if pending is not None:
+        raise SimpleHashPendingEventFound(f"Event Identity {pending}")
+
+
+def canonicalize_event(event):
+    """Canonicalize a single event"""
+    missing = set(V1_FIELDS).difference(event)
+    if missing:
+        raise SimpleHashFieldMissing(
+            f"Event Identity {event['identity']} has missing field(s) {missing}"
+        )
+    return {k: event[k] for k in V1_FIELDS}
+
+
+def canonicalize_events(events):
+    """Form list of events only containing necessary fields"""
+    return [
+        canonicalize_event(event)
+        for event in sorted(events, key=itemgetter("identity"))
+    ]
+
+
+def hash_events(events):
+    """Generate Simplehash for a given set of events canonicalizing then hashing"""
+
+    check_for_pending_events(events)
+    canonical_events = canonicalize_events(events)
+
+    # bencode the events
+    bencoded_events = binary_encode(canonical_events)
+
+    # Hash the bencoded events
+    events_hash = sha256(bencoded_events).hexdigest()
+
+    return events_hash
+
+
+def main():
+    """Reads from stdin"""
+
+    events_hash = hash_events(json_load(sys_stdin))
```

Comment:
What do we want the input to this to be?

Do we want a list of events,
Or do we want to pass in a start and end range?
Do we want the fully qualified url that lists the events?

Reply:
good question - look at REAME,rst